src/commonMain/kotlin/io/ktor/io/ByteArrayBuffer.kt

```diff
@@ -1,184 +1,115 @@
+        checkCanRead(index, 2, writeIndex)
+        return array.loadShortAt(index)
     }
 
-    override fun readShort(): Short {
-        checkHasBytesToRead(2)
-        return doReadShort()
+    override fun loadIntAt(index: Int): Int {
+        checkCanRead(index, 4, writeIndex)
+        return array.loadIntAt(index)
     }
 
-    override fun readInt(): Int {
-        checkHasBytesToRead(4)
-        return doReadInt()
+    override fun loadLongAt(index: Int): Long {
+        checkCanRead(index, 8, writeIndex)
+        return array.loadLongAt(index)
     }
 
-    override fun readLong(): Long {
-        checkHasBytesToRead(8)
-        return doReadLong()
+    override fun storeByteAt(index: Int, value: Byte) {
+        checkCanWrite(index, 1, capacity)
+        array[index] = value
     }
 
-    override fun writeByte(value: Byte) {
-        checkHasSpaceToWrite(1)
-        doWriteByte(value)
+    override fun storeShortAt(index: Int, value: Short) {
+        checkCanWrite(index, 2, capacity)
+        array.storeShortAt(index, value)
     }
 
-    override fun writeShort(value: Short) {
-        checkHasSpaceToWrite(2)
-        doWriteShort(value)
+    override fun storeIntAt(index: Int, value: Int) {
+        checkCanWrite(index, 4, capacity)
+        array.storeIntAt(index, value)
     }
 
-    override fun writeInt(value: Int) {
-        checkHasSpaceToWrite(4)
-        doWriteInt(value)
+    override fun storeLongAt(index: Int, value: Long) {
+        checkCanWrite(index, 8, capacity)
+        array.storeLongAt(index, value)
     }
 
-    override fun writeLong(value: Long) {
-        checkHasSpaceToWrite(8)
-        doWriteLong(value)
+    override fun storeBufferAt(index: Int, value: Buffer): Int {
+        return value.copyToArray(array, index, capacity)
     }
 
-    override fun read(destination: ByteArray, startIndex: Int, endIndex: Int): Int {
-        checkArrayBounds(destination, startIndex, endIndex)
-
-        val count = min(endIndex - startIndex, writeIndex - readIndex)
-        array.copyInto(destination, startIndex, readIndex, readIndex + count)
-        readIndex += count
+    override fun storeArrayAt(index: Int, value: ByteArray, startPosition: Int, endPosition: Int): Int {
+        require(startPosition >= 0) { "startPosition($startPosition) must be >= 0" }
+        require(endPosition <= value.size) { "endPosition($endPosition) must be <= value.size(${value.size})" }
+        require(startPosition <= endPosition) { "startPosition($startPosition) must be <= endPosition($endPosition)" }
 
+        val count = min(capacity - index, endPosition - startPosition)
+        value.copyInto(array, index, startPosition, startPosition + count)
         return count
     }
 
-    override fun write(source: ByteArray, startIndex: Int, endIndex: Int): Int {
-        checkArrayBounds(source, startIndex, endIndex)
+    override fun readArray(): ByteArray = array.sliceArray(readIndex until writeIndex)
```

Comment:
will `sliceArray` be optimized by compiler, or will create range on every call?
if second, than it will be better to change it to `copyOfRange`

Reply:
dropped readArray